File: examples/simple-examples-java/src/main/java/ai/koog/agents/example/calculator/CalculatorTools.java
package ai.koog.agents.example.calculator;

import ai.koog.agents.core.tools.annotations.LLMDescription;
import ai.koog.agents.core.tools.annotations.Tool;
import ai.koog.agents.core.tools.reflect.ToolSet;

/**
 * Java implementation of calculator tools, mirroring {@code CalculatorTools.kt}.
 *
 * <p>Demonstrates how to define agent tools using the {@code @Tool} and {@code @LLMDescription}
 * annotations on a class that implements {@link ToolSet}.
 *
 * <p>Note: {@code @LLMDescription} uses {@code value} as its annotation attribute.
 */
@LLMDescription("Tools for basic calculator operations")
public class CalculatorTools implements ToolSet {

    @Tool
    @LLMDescription("Adds two numbers")
    public String plus(
        @LLMDescription("First number") float a,
        @LLMDescription("Second number") float b
    ) {
        return String.valueOf(a + b);
    }

    @Tool
    @LLMDescription("Subtracts the second number from the first")
    public String minus(
        @LLMDescription("First number") float a,
        @LLMDescription("Second number") float b
    ) {
        return String.valueOf(a - b);
    }

    @Tool
    @LLMDescription("Divides the first number by the second")
    public String divide(
        @LLMDescription("First number") float a,
        @LLMDescription("Second number") float b
    ) {
        return String.valueOf(a / b);
    }

    @Tool
    @LLMDescription("Multiplies two numbers")
    public String multiply(
        @LLMDescription("First number") float a,
        @LLMDescription("Second number") float b
    ) {
        return String.valueOf(a * b);
    }
}

File: examples/simple-examples/src/main/kotlin/ai/koog/agents/example/calculator/CalculatorTools.kt
package ai.koog.agents.example.calculator

import ai.koog.agents.core.dsl.builder.strategy
import ai.koog.agents.core.dsl.extension.ReceivedToolResults
import ai.koog.agents.core.dsl.extension.nodeExecuteTools
import ai.koog.agents.core.dsl.extension.nodeLLMCompressHistory
import ai.koog.agents.core.dsl.extension.nodeLLMRequest
import ai.koog.agents.core.dsl.extension.nodeLLMSendToolResults
import ai.koog.agents.core.dsl.extension.onTextMessage
import ai.koog.agents.core.dsl.extension.onToolCalls
import ai.koog.agents.core.tools.annotations.LLMDescription
import ai.koog.agents.core.tools.annotations.Tool
import ai.koog.agents.core.tools.reflect.ToolSet

@Suppress("unused")
@LLMDescription("Tools for basic calculator operations")
class CalculatorTools : ToolSet {

    @Tool
    @LLMDescription("Adds two numbers")
    fun plus(
        @LLMDescription("First number")
        a: Float,

        @LLMDescription("Second number")
        b: Float
    ): String {
        return (a + b).toString()
    }

    @Tool
    @LLMDescription("Subtracts the second number from the first")
    fun minus(
        @LLMDescription("First number")
        a: Float,

        @LLMDescription("Second number")
        b: Float
    ): String {
        return (a - b).toString()
    }

    @Tool
    @LLMDescription("Divides the first number by the second")
    fun divide(
        @LLMDescription("First number")
        a: Float,

        @LLMDescription("Second number")
        b: Float
    ): String {
        return (a / b).toString()
    }

    @Tool
    @LLMDescription("Multiplies two numbers")
    fun multiply(
        @LLMDescription("First number")
        a: Float,

        @LLMDescription("Second number")
        b: Float
    ): String {
        return (a * b).toString()
    }
}

object CalculatorStrategy {
    private const val MAX_TOKENS_THRESHOLD = 1000

    val strategy = strategy<String, String>("test") {
        val nodeCallLLM by nodeLLMRequest()
        val nodeExecuteToolMultiple by nodeExecuteTools(parallel = true)
        val nodeSendToolResultMultiple by nodeLLMSendToolResults()
        val nodeCompressHistory by nodeLLMCompressHistory<ReceivedToolResults>()

        edge(nodeStart forwardTo nodeCallLLM)

        edge(
            (nodeCallLLM forwardTo nodeFinish)
                onTextMessage { true }
        )

        edge(
            (nodeCallLLM forwardTo nodeExecuteToolMultiple)
                onToolCalls { true }
        )

        edge(
            (nodeExecuteToolMultiple forwardTo nodeCompressHistory)
                onCondition { llm.readSession { prompt.latestTokenUsage > MAX_TOKENS_THRESHOLD } }
        )

        edge(nodeCompressHistory forwardTo nodeSendToolResultMultiple)

        edge(
            (nodeExecuteToolMultiple forwardTo nodeSendToolResultMultiple)
                onCondition { llm.readSession { prompt.latestTokenUsage <= MAX_TOKENS_THRESHOLD } }
        )

        edge(
            (nodeSendToolResultMultiple forwardTo nodeExecuteToolMultiple)
                onToolCalls { true }
        )

        edge(
            (nodeSendToolResultMultiple forwardTo nodeFinish)
                onTextMessage { true }
        )
    }
}
